src/core/server/services/spam/index.ts

```diff
@@ -0,0 +1,171 @@
+import { User } from "@sentry/node";
+import { Db } from "mongodb";
+import fetch from "node-fetch";
+
+import logger from "coral-server/logger";
+import { Comment, getLatestRevision } from "coral-server/models/comment";
+import { retrieveStory, Story } from "coral-server/models/story";
+import { Tenant } from "coral-server/models/tenant";
+import { retrieveUser } from "coral-server/models/user";
+import { Request } from "coral-server/types/express";
+
+export interface Parameters {
+  apiKey: string;
+  comment: Readonly<Comment>;
+  story: Readonly<Story>;
+  author: Readonly<User>;
+  userIP: string | undefined;
+  userAgent: string | undefined;
+  referrer: string | undefined;
+}
+
+const createBody = (params: Parameters) => {
+  const { comment, story, author, userIP, userAgent, referrer } = params;
+  const latestRevision = getLatestRevision(comment);
+  const username = author ? (author.username ? author.username : "") : "";
+
+  return {
+    user_ip: userIP,
+    user_agent: userAgent,
+    referrer,
+
+    permalink: story.url,
+
+    comment_type: "comment",
+    comment_author: username,
+    comment_content: latestRevision ? latestRevision.body : "",
+    comment_date_gmt: latestRevision
+      ? latestRevision.createdAt.toISOString()
+      : "",
+
+    is_test: false,
+  };
+};
+
+const submitSpam = async (params: Parameters) => {
+  const { apiKey } = params;
+  const url = `https://${apiKey}.rest.akismet.com/1.1/submit-spam`;
+  const body = createBody(params);
+
+  try {
+    const response = await fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(body),
+    });
+
+    return response.ok;
+  } catch (err) {
+    logger.error({ err }, "unable to submit spam to Akismet");
+    return false;
+  }
+};
+
+const submitNotSpam = async (params: Parameters) => {
+  const { apiKey } = params;
+  const url = `https://${apiKey}.rest.akismet.com/1.1/submit-ham`;
+  const body = createBody(params);
+
+  try {
+    const response = await fetch(url, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(body),
+    });
+
+    return response.ok;
+  } catch (err) {
+    logger.error({ err }, "unable to submit non-spam to Akismet");
+    return false;
+  }
+};
+
+export const submitCommentAsNotSpam = async (
+  mongo: Db,
+  tenant: Tenant,
+  comment: Readonly<Comment>,
+  request?: Request | undefined
+) => {
+  const story = await retrieveStory(mongo, tenant.id, comment.storyID);
+  if (!story) {
+    return;
+  }
+
+  const author = comment.authorID
+    ? await retrieveUser(mongo, tenant.id, comment.authorID)
+    : null;
+  if (!author) {
+    return;
+  }
+
+  const userIP = request?.ip || "";
+
+  const userAgent = request?.get("User-Agent");
+  if (!userAgent || userAgent.length === 0) {
+    return;
```

Comment:
are user agent and referrer required by akismet?

Reply:
As far as I know, they are. I was going off our existing `src/core/server/services/comments/pipeline/phases/spam.ts` API calls:

```
const isSpam = await client.checkSpam({
      user_ip: userIP, // REQUIRED
      referrer, // REQUIRED
      user_agent: userAgent, // REQUIRED
      comment_content: comment.body,
      permalink: story.url,
      comment_author: author.username || "",
      comment_type: "comment",
      is_test: false,
    });
```